src/solver.hs

```diff
@@ -0,0 +1,232 @@
+module Solver where
+
+import Sudoku
+import SudokuLogic
+import Sudokus
+import Optimizer
+import Data.List
+data Step =     LastFreeCellBlock Position      |
+                LastFreeCellRow Position        |
+                LastFreeCellCollumn Position    |
+                SingleCandidate Position        |
+                SinglePositionRow Position      |
+                SinglePositionColumn Position   |
+                SinglePositionBlock Position    |
+                CandidateLine Position          |
+                NakedPairRow Position           |
+                NakedPairColumn Position        |
+                NakedPairBlock Position         |
+                HiddenPairRow Position          |
+                HiddenPairColumn Position       |
+                HiddenPairBlock Position        |
+                NOAVAILABLESTEPS
+    deriving (Eq,Show) -- | Then add more step types
+
+type Solution = [Step]
+
+
+-- | Takes a sudoku and returns a reorginised list of steps
+optimSteps :: Sudoku -> [Position -> Step]
+optimSteps sud = sudStepOrder $ zip (stepWeight sud) steps
+
+-- | List of every step the solver can try
+steps :: [Position -> Step]
+steps = [--LastFreeCellBlock, LastFreeCellRow, LastFreeCellCollumn, 
+        SingleCandidate,
+        SinglePositionRow, SinglePositionColumn, SinglePositionBlock
+        ,NakedPairBlock, NakedPairRow, NakedPairColumn
+        ,HiddenPairBlock, HiddenPairRow, HiddenPairColumn
+        ,CandidateLine
+        ]
+
```

Comment:
You treat every technique as if it was specified by just a position.

But actually, many techniques take more parameters as input, for example which row/column/blocks you are working with, or with a pair of cells, not just one. The current way, the technique has to reconstruct all that information and there is a lot of duplication.

Reply:
This information is also needed if you want to produce meaningful output, like: apply technique X in row Y with block Z.